proposals/Reliable Mutability Primitive.md

```diff
@@ -0,0 +1,166 @@
+
+
+
+### Project definition
+##### Brief plan of attack
+
+<!--Briefly describe the milestones/steps/work needed for this project-->
+
+- Implement third party IPNS republishing (TODO: Find a relevant issue link)
+- Collaborate with Pinning Services to make this part of the API.
+- Fix IPNS publishing in browsers.
+- Enable pubsub by default + IPNS over pubsub.
+- Implement & deploy strategic republishing system on PL hosted nodes that will republish popular content and drop stalled one. So that IPNS works even thought pinning services.
+- (Nice to have) Implement and rollout delegated publishing to enable updating from multiple devices.
+- (Nice to have) Implement and rollout versioning system so that nodes can see full update history.
+
+##### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+Publishing IPNS record takes just a few miliseconds. Shutting laptop and loading `https://ipfs.io/ipns/{key}` week later on other device loads the content published as fast as loading corresponding `https://ipfs.io/ipfs/{cid}`.
+
+
+#####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+<!--
+Provide success criteria. These might include particular metrics, desired changes in the types of bug reports being filed, desired changes in qualitative user feedback (measured via surveys, etc), etc.
+-->
+People and teams are storing IPNS keys in [DNSLink][] records as opposed to CIDs, because that provides a more effective and simpler way to publish updates. We see new projects are leveraging IPNS over working around it via things like [DNSLink][].
+
+
+##### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+- IPNS has bad reputation, changing it may prove to be a challenge.
+- Teams may find that general primitive is not adequate for custom needs.
+- Required key management (e.g. key recovery, key compromise) still may prove it impractical.
+- Higher level alternatives like textile threads may prove to be better layer of abstraction.
+
+##### Alternatives
+_How might this project’s intent be realized in other ways (other than this project proposal)? What other potential solutions can address the same need?_
+
+- Integrating textile threads as core component of the ecosystem.
+- Standardizing DNS record updates mechanism and intergating into IPFS.
+- Work with Pinning Services to implement stardized DNSLink updating mechanism.
+
+##### Dependencies/prerequisites
+<!--List any other projects that are dependencies/prerequisites for this project that is being pitched.-->
+
+- Functioning IPNS in browsers
```

Comment:
Note: likely includes an up-to-date IPNS over PubSub implementation

Reply:
@aschmahmann do you mean something like 

```suggestion
- Functioning IPNS in browsers (over up-to-date PubSub implementation)
```

Or a separate item ? I

Reply:
Well IPNS over PubSub has a few tweaks over just PubSub itself to give some reliability. It's specced [here](https://github.com/ipfs/specs/blob/a370a6aa2345b251a4c04d37b8c50d3aa8f2701f/naming/pubsub.md).

@hugomrdias would probably know more about other work required to get IPNS functioning in a browser I just know that without the IPNS over PubSub changes in the spec that it won't be fast enough.

Reply:
Sorry for the basic question, but I'm still confused by what is meant by "functioning IPNS in browsers".  What isn't working browsers today?

Reply:
I should probably capture this in the document. Currently pubsub is disabled in go-ipfs so all the IPNS records are published and resolved from DHT. Nodes in browsers do not have necessary capabilities to query/publish DHT. Instead they do publish and resolve IPNS records over pubsub. Result is fragmented network where go and js nodes see each others records.

Furthermore because because there is not enough nodes on the network with pubsub enabled published records from browser don't always make to nodes subscribed to them.

P.S.: Some of this may not be completely accurate, @hugomrdias @aschmahmann could you please confirm I'm not misrepresenting current state.